src/Sarif/Baseline/V2/V2ResultMatcher.cs

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.CodeAnalysis.Sarif.Baseline.ResultMatching;
+
+namespace Microsoft.CodeAnalysis.Sarif.Baseline
+{
+    public class V2ResultMatcher : IResultMatcher
+    {
+        public IList<MatchedResults> Match(IList<ExtractedResult> before, IList<ExtractedResult> after)
+        {
+            ResultMatchState state = new ResultMatchState(before, after);
+            return state.Match();
+        }
+    }
+
+    /// <summary>
+    ///  ResultMatchState contains all of the state needed to compute matches
+    ///  between two batches of Results.
+    /// </summary>
+    internal class ResultMatchState
+    {
+        // Results in the first set
+        public List<ExtractedResult> Before;
+
+        // Results in the second set
+        public List<ExtractedResult> After;
+
+        // The index of the Result in After matching this Result in Before, or -1
+        public int[] MatchingIndexFromBefore;
+
+        // The index of the Result in Before matching this Result in After, or -1
+        public int[] MatchingIndexFromAfter;
+
+        public ResultMatchState(IList<ExtractedResult> before, IList<ExtractedResult> after)
+        {
+            // Sort results by 'Where' for matching
+            Before = new List<ExtractedResult>(before);
+            Before.Sort(WhereComparer.Instance);
+
+            After = new List<ExtractedResult>(after);
+            After.Sort(WhereComparer.Instance);
+
+            // Set all match indices to -1 initially (no Results matched)
+            MatchingIndexFromBefore = new int[Before.Count];
+            Fill(MatchingIndexFromBefore, -1);
+
+            MatchingIndexFromAfter = new int[After.Count];
+            Fill(MatchingIndexFromAfter, -1);
+        }
+
+        public IList<MatchedResults> Match()
+        {
+            if (Before.Count == 1 && After.Count == 1)
```

Comment:
>if [](start = 12, length = 2)

Is this common enough to special case?

Reply:
Just one result is pretty common in Nathan's examples.
The issue is that if there's only one result, the "match before" and "match after" logic doesn't fire for them, so this is the only way they'll actually get compared.

Reply:
I added a comment explaining this.

---
In reply to: [295892427](https://github.com/microsoft/sarif-sdk/pull/1539#discussion_r295892427) [](ancestors = 295892427)